cmd/threatest/run.go

```diff
@@ -0,0 +1,221 @@
+		if err := m.writeJsonOutput(results); err != nil {
+			return err
+		}
+		log.Infof("Wrote scenario test results to %s", m.JsonOutputFile)
+	}
+
+	// Return an error to exit with a non-zero status code if at least one test failed
+	if hasError {
+		return fmt.Errorf("at least 1 scenario failed")
+	} else {
+		return nil
+	}
+}
+
+func (m *RunCommand) Validate() error {
+	if len(m.InputFiles) == 0 {
+		return errors.New("please provide at least 1 scenario")
+	}
+
+	// If an SSH key is provided, check it exists
+	if sshKey := m.SSHConfig.SSHKey; sshKey != "" {
+		if _, err := os.Stat(sshKey); err != nil && sshKey != "" {
+			return fmt.Errorf("invalid SSH key file %s: %v", sshKey, err)
+		}
+	}
+
+	return nil
+}
+
+// runScenariosParallel runs all the provided scenarios in parallel, honoring the maximum parallelism
+// every time a test completes, the callback function is invoked
+func (m *RunCommand) runScenariosParallel(allScenarios []*threatest.Scenario, callback func(result *ScenarioRunResult)) []ScenarioRunResult {
+	numWorkers := m.Parallelism
+	// No point in having more workers than scenarios to run
+	if numScenarios := len(allScenarios); numScenarios < numWorkers {
+		numWorkers = numScenarios
+	}
+
+	log.Infof("Running %d scenarios with a parallelism of %d", len(allScenarios), numWorkers)
+
+	// Channel to hold "tasks" (scenarios to run)
+	scenarioChan := make(chan *threatest.Scenario, numWorkers)
+	// Channel to hold test results
+	resultsChan := make(chan *ScenarioRunResult)
+
+	// Create 1 worker by desired parallelism unit
+	for worker := 0; worker < numWorkers; worker++ {
+		go m.runSingleScenario(scenarioChan, resultsChan)
+	}
+
+	// Submit each scenario
+	for _, scenario := range allScenarios {
+		scenarioChan <- scenario
+	}
+
+	// Retrieve results as they are produced
+	var allResults []ScenarioRunResult
+	for range allScenarios {
+		result := <-resultsChan
+		callback(result)
+		allResults = append(allResults, *result)
+	}
+
+	return allResults
+}
+
+// runSingleScenario runs inside a goroutine and uses Threatest to run one scenario
+func (m *RunCommand) runSingleScenario(scenarios <-chan *threatest.Scenario, results chan<- *ScenarioRunResult) {
+	for scenario := range scenarios {
+		runner := threatest.Threatest()
+		runner.Scenarios = append(runner.Scenarios, scenario)
+		runner.Interval = 2 * time.Second
+
+		start := time.Now()
+		err := runner.Run()
+		end := time.Now()
+
+		var errorMessage = ""
+		if err != nil {
+			errorMessage = err.Error()
+		}
+
+		results <- &ScenarioRunResult{
+			Description:     scenario.Name,
+			ErrorMessage:    errorMessage,
+			Success:         err == nil,
+			DurationSeconds: end.Sub(start).Seconds(),
+			TimeDetonated:   start,
+		}
+	}
+}
+
+func (m *RunCommand) writeJsonOutput(results []ScenarioRunResult) error {
+	outputBytes, err := json.MarshalIndent(results, "", "  ")
+	if err != nil {
+		return fmt.Errorf("unable to convert scenario test results to JSON: %v", err)
```

Comment:
[nit] Go supports a `%w` format instruction which does the same thing as `%v` with the following extra behaviour:
> In Go 1.13, the fmt.Errorf function supports a new %w verb. When this verb is present, the error returned by fmt.Errorf will have an Unwrap method returning the argument of %w, which must be an error. In all other ways, %w is identical to %v.

(taken from https://go.dev/blog/go1.13-errors)

Reply:
interesting, thanks! will add

Reply:
done in 2e80cc816cc5b7f0ab091cbf06e6850edb6ce1a1